AsyncSnapshot.data to throw if error or no data

tvolkert opened the pull request:
## Description

Currently, if you attempt to access the `AsyncSnapshot.data` property when the error property is non-null, it will return null.  This API often confuses users, who expect the property to throw an error if the error property is set.

This change updates the API to throw an error if `error` is set -- and to correspondingly remove the `requireData` property.

After receiving this update, callers should update their code accordingly:
* If callers were accessing the `data` property without first checking `hasData` or `hasError`, they should update their code to check `hasData` or `hasError`, or to wrap their `data` access in a try/catch.
* If callers were using the `requireData` property, they should update their code to simply access the `data` property instead.

## Related Issues

https://github.com/flutter/flutter/issues/34545

## Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [x] Yes, this is a breaking change.  The following announcement was sent to flutter-announce: https://groups.google.com/forum/#!topic/flutter-announce/H6Od0QdsdrI
- [ ] No, this is *not* a breaking change.

a14n reviewed (approved):
LGTM

goderbauer commented on packages/flutter/lib/src/widgets/async.dart:
```diff
@@ -182,6 +184,21 @@ enum ConnectionState {
+/// reference to this object, thus adhering to the contract of an `@immutable`
+/// class (at least technically, although not in spirit).
+///
+/// This class is temporary and only exists to facilitate a soft transition
```
I am not sure I totally understand the transition path: If the ultimate goal is to make "data" throw if an error is present, shouldn't we for now just fire an assert if "data" is accessed while an error is present?

goderbauer commented on packages/flutter/lib/src/widgets/async.dart:
```diff
@@ -182,6 +184,21 @@ enum ConnectionState {
+///
+/// The level of indirection allows [AsyncSnapshot] to maintain a `final`
+/// reference to this object, thus adhering to the contract of an `@immutable`
+/// class (at least technically, although not in spirit).
```
What do we win by stying immutable as a technicality, but not in spirit?

goderbauer commented on packages/flutter/lib/src/widgets/async.dart:
```diff
@@ -220,17 +239,35 @@ class AsyncSnapshot<T> {
   /// set to an initial data value specified by the relevant widget. See
   /// [FutureBuilder.initialData] and [StreamBuilder.initialData].
-  final T data;
+  T get data {
```
The doc comment above needs updating to reflect the changed behavior.

goderbauer replied:
(And it might be worth to describe the general behavior in the doc comment of the class as well)

tvolkert commented:
@goderbauer I made the changes we discussed:

1. Add a new `value` getter that throws if there's no value (as `requireData` currently does)
2. Add a corresponding new `hasValue` getter
3. Deprecate both `data` and `hasData`
4. Remove `requireData`

I don't love it, because `value` is not as good a name as `data`.  `data` matches `Stream.listen`'s `onData` callback -- and it also exists in `StreamBuilder.initialData`.

Then again, I think the existing API is broken, so.... wdyt?

goderbauer reviewed (commented):
It is sad that we lose the similarity to the Stream/StreamBuilder API. In the interest of a clean API, maybe we should just do the full-on breaking change under the existing `data` name?

tvolkert committed:
AsyncSnapshot.data to throw if error or no data

This updates `AsyncSnapshot.data` to act as `AsyncSnapshot.requireData`
used to -- and it removes `AsyncSnapshot.requireData`.

https://github.com/flutter/flutter/issues/34545
https://groups.google.com/forum/#!topic/flutter-announce/H6Od0QdsdrI

tvolkert commented:
Ok, @goderbauer I made the changes as discussed and sent out the breaking change announcement.  PTAL.

goderbauer commented:
(PR triage): This will need some changes to address the concerns raised in https://github.com/flutter/flutter/issues/34545.

tvolkert commented:
Ok, updates per the conversation in #34545 have all been made.  This is ready for another review.

/cc @rrousselGit

tvolkert committed:
Fixed device lab test

tvolkert committed:
Another devicelab fix

a14n commented on packages/flutter/lib/src/widgets/async.dart:
```diff
@@ -193,46 +201,61 @@ enum ConnectionState {
-  /// and optionally either [data] or [error] (but not both).
-  const AsyncSnapshot._(this.connectionState, this.data, this.error)
+  /// Creates an [AsyncSnapshot] with the specified [connectionState] and
+  /// [hasData], and optionally either [data] or [error] (but not both).
```
should it be `[_data]` as it refers to parameter here?

tvolkert replied:
No, I'm linking to the property intentionally.

https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#dartdoc-specific-requirements

a14n commented on packages/flutter/lib/src/widgets/async.dart:
```diff
@@ -193,46 +201,61 @@ enum ConnectionState {
+  /// Creates an [AsyncSnapshot] with the specified [connectionState] and
+  /// [hasData], and optionally either [data] or [error] (but not both).
+  ///
+  /// It is legal for both [hasData] to be true and [data] to be null.
```
ditto

tvolkert replied:
Same here

tvolkert committed:
Fixed analysis of sample code

Hixie commented on packages/flutter/lib/src/widgets/async.dart:
```diff
@@ -193,46 +201,61 @@ enum ConnectionState {
 
-  /// Current state of connection to the asynchronous computation.
+  /// The current state of the connection to the asynchronous computation.
   final ConnectionState connectionState;
```
needs more docs

Hixie commented on packages/flutter/lib/src/widgets/async.dart:
```diff
@@ -193,46 +201,61 @@ enum ConnectionState {
-      return data;
+      return _data;
     if (hasError)
       throw error;
```
pity we can't preserve the stack trace

Hixie replied:
maybe add a TODO pointing at https://github.com/dart-lang/sdk/issues/30741.

Hixie commented on packages/flutter/lib/src/widgets/async.dart:
```diff
@@ -243,26 +266,20 @@ class AsyncSnapshot<T> {
 
   @override
-  String toString() => '$runtimeType($connectionState, $data, $error)';
+  String toString() => '$runtimeType($connectionState, $hasData, $_data, $error)';
```
hasData is a bool, so that will be confusing. maybe `${hasData ? "has data" : "no data"}` ?

Hixie replied:
also, probably omit `$_data` if there's no data, and omit error if there's no error? in general this could be clearer in various ways.

Hixie commented on packages/flutter/lib/src/widgets/async.dart:
```diff
@@ -272,12 +289,13 @@ class AsyncSnapshot<T> {
       return false;
```
should use runtimeType not `is`

Hixie commented on packages/flutter/lib/src/widgets/async.dart:
```diff
@@ -372,42 +403,87 @@ typedef AsyncWidgetBuilder<T> = Widget Function(BuildContext context, AsyncSnaps
-  /// The [initialData] is used to create the initial snapshot.
+  /// The [initialData] argument is used to create the initial snapshot.
   ///
   /// The [builder] must not be null.
```
docs should talk about what to do if you don't have the initial data (the common case)

Hixie commented on packages/flutter/lib/src/widgets/async.dart:
```diff
@@ -329,15 +347,27 @@ typedef AsyncWidgetBuilder<T> = Widget Function(BuildContext context, AsyncSnaps
-/// to be processed.
+/// to be processed. In cases where callers wish to have no initial data, the
+/// [new StreamBuilder.withoutInitialData] constructor may be used. Doing so
+/// may cause the first frame to have a snapshot that contains no data.
```
may? definitely will, no?

tvolkert replied:
It's possible the stream will have produced data by the first frame. The docs above specifically call out that since the builder is called at the discretion of the Flutter pipeline, callers are not guaranteed to see every snapshot in rendered frames.

Hixie commented on packages/flutter/lib/src/widgets/async.dart:
```diff
@@ -372,42 +403,87 @@ typedef AsyncWidgetBuilder<T> = Widget Function(BuildContext context, AsyncSnaps
+  /// access the [initialData] property will throw a [StateError].
+  T get initialData {
+    if (!hasInitialData)
+      throw StateError('No initial data');
```
can we have a more useful error message?

tvolkert replied:
Would it be more appropriate to use `FlutterError` here, or is `StateError` better?  The docs on `FlutterError` say "Error class used to report Flutter-specific assertion failures and contract violations," but it's not clear to me from that description if widget implementations are encouraged to throw `FlutterError` or not.

Hixie commented on packages/flutter/lib/src/widgets/async.dart:
```diff
@@ -372,42 +403,87 @@ typedef AsyncWidgetBuilder<T> = Widget Function(BuildContext context, AsyncSnaps
   @override
   AsyncSnapshot<T> afterData(AsyncSnapshot<T> current, T data) {
-    return AsyncSnapshot<T>.withData(ConnectionState.active, data);
+    return _TypeLiteral.isVoidType(T)
```
it's weird that we special-case `void`, though I guess it's understandable.

Hixie replied:
we should definitely document this though

Hixie replied:
(in the class docs, e.g.)

tvolkert replied:
I already do - see the docs lines starting with "The `StreamBuilder<void>` and `StreamBuilder<Null>` types will produce"

Hixie commented on packages/flutter/lib/src/widgets/async.dart:
````diff
@@ -498,6 +588,7 @@ class StreamBuilder<T> extends StreamBuilderBase<T, AsyncSnapshot<T>> {
 /// ```dart
 /// FutureBuilder<String>(
 ///   future: _calculation, // a previously-obtained Future<String> or null
````
we should make it clearer that _calculation _must not be created during the call to the build method_ that creates the `FutureBuilder`.

Hixie commented on packages/flutter/lib/src/widgets/async.dart:
```diff
@@ -520,20 +611,44 @@ class FutureBuilder<T> extends StatefulWidget {
+  /// [future] has completed will contain no data.
+  ///
+  /// The [builder] argument must not be null.
+  const FutureBuilder.withoutInitialData({
```
isn't this the common case? or even the only case, usually? seems weird to make this harder to use.

Hixie commented on packages/flutter/lib/src/widgets/async.dart:
```diff
@@ -542,30 +657,68 @@ class FutureBuilder<T> extends StatefulWidget {
+  /// If [hasInitialData] is false, this will throw a [StateError].
+  T get initialData {
+    if (!hasInitialData)
+      throw StateError('No initial data');
```
as above, assuming we keep this feature at all. it's weird to me to have "initial data" for a future.

Hixie commented on packages/flutter/lib/src/widgets/async.dart:
```diff
@@ -632,3 +789,39 @@ class _FutureBuilderState<T> extends State<FutureBuilder<T>> {
+  /// Currently in the Dart language, both the `Null` type and the `void` type
+  /// represents "void" types (types that contain no value).
+  static bool isVoidType(Type type) {
+    return type == const _TypeLiteral<Null>().type  // ignore: prefer_void_to_null
```
i wouldn't accept Null here. If it's a `Future<Null>` and it completes (with null) then I do have data: null.

tvolkert committed:
Review comments

tvolkert commented:
@Hixie all comments addressed - PTAL

tvolkert committed:
Fix tests

Hixie commented on packages/flutter/lib/src/widgets/async.dart:
```diff
@@ -520,11 +628,21 @@ class FutureBuilder<T> extends StatefulWidget {
     this.future,
-    this.initialData,
+    @Deprecated(
+      "Don't provide initialData to FutureBuilder. Instead, check for "
```
we generally would phrase this more softly as "Instead of providing initialData to a FutureBuilder, consider checking for..."

Hixie commented on packages/flutter/lib/src/widgets/async.dart:
```diff
@@ -542,29 +663,53 @@ class FutureBuilder<T> extends StatefulWidget {
+  /// error itself will be available in [AsyncSnapshot.error], and
   /// [AsyncSnapshot.hasError] will be true.)
+  @Deprecated(
+    "Don't use FutureBuilder.initialData. Instead, check for "
```
similar here

Hixie commented on packages/flutter/lib/src/widgets/async.dart:
```diff
@@ -542,29 +663,53 @@ class FutureBuilder<T> extends StatefulWidget {
+    "Don't use FutureBuilder.initialData. Instead, check for "
+    'ConnectionState.none or ConnectionState.waiting in your build() '
+    'method to know whether the future has completed or not.',
+  )
```
deprecation explanation should be in `///` dartdocs too

Hixie commented:
Mentions of `initialData` elsewhere in the docs should also discourage it.
Constructor docs should briefly reiterate the recommendation to avoid creating a new future during build.

LGTM.

tvolkert committed:
Review comments

creativecreatorormaybenot commented:
Is there a documentation for the `StreamBuilder` changes? I am expecting a lot of users to be confused.

I guess I will be answering some questions on StackOverflow soon 🤷

tvolkert commented:
@creativecreatorormaybenot I also added a section to the changelog that covers the changes as well, with examples.  https://github.com/flutter/flutter/wiki/Changelog#v183

creativecreatorormaybenot commented:
@tvolkert Awesome, that's what I was missing previously.

amirh commented:
What's the migration plan for a package that needs to work both on Flutter master and stable, and is broken by this change?

tvolkert commented:
I actually just sent out https://github.com/flutter/flutter/pull/36577, which would swap the semantics of the two `StreamBuilder` constructors, allowing us a deprecation path rather than a hard API break.

amirh commented:
Won't that shift the problem(of not being able to work on both master and stable) to packages that provide initialData?

amirh commented:
take that back - I thought that PR removed `initialData` which isn't the case.

tvolkert commented:
> What's the migration plan for a package that needs to work both on Flutter master and stable, and is broken by this change?

@amirh there's still discussion in https://github.com/flutter/flutter/issues/34545#issuecomment-513446461 that could mean we don't merge #36577.  If that ends up being the best course of action, then the migration plan for packages that need to work on both master and stable is to have them explicitly pass `initialData: null` to the `StreamBuilder` constructor, which will yield the same behavior that they used to get.
